File: sniffer_tool/src/packet_analyzer.py
from scapy.all import IP, TCP, UDP
from colorama import Fore, Style
from datetime import datetime

class PacketAnalyzer:
    @staticmethod
    def analyze_packet(packet):
        if not packet.haslayer(IP):
            return None

        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
        ip_packet = packet[IP]
        
        # Basic packet info
        packet_info = {
            "timestamp": timestamp,
            "source_ip": ip_packet.src,
            "dest_ip": ip_packet.dst,
            "protocol": ip_packet.proto,
            "length": len(packet),
        }

        # Add TCP-specific info
        if packet.haslayer(TCP):
            tcp = packet[TCP]
            packet_info.update({
                "source_port": tcp.sport,
                "dest_port": tcp.dport,
                "flags": tcp.flags
            })
        
        # Add UDP-specific info
        elif packet.haslayer(UDP):
            udp = packet[UDP]
            packet_info.update({
                "source_port": udp.sport,
                "dest_port": udp.dport
            })

        return packet_info

    @staticmethod
    def format_packet_info(packet_info):
        if not packet_info:
            return ""

        protocol_map = {6: "TCP", 17: "UDP"}
        protocol = protocol_map.get(packet_info["protocol"], str(packet_info["protocol"]))
        
        output = (
            f"{Fore.YELLOW}{packet_info['timestamp']}{Style.RESET_ALL} | "
            f"{Fore.GREEN}{packet_info['source_ip']}:{packet_info.get('source_port', '?')}{Style.RESET_ALL} → "
            f"{Fore.RED}{packet_info['dest_ip']}:{packet_info.get('dest_port', '?')}{Style.RESET_ALL} | "
            f"{Fore.BLUE}{protocol}{Style.RESET_ALL} | "
            f"Length: {packet_info['length']} bytes"
        )

        if "flags" in packet_info:
            output += f" | Flags: {packet_info['flags']}"

        return output

File: src/packet-analyzer.js
const chalk = require('chalk');

class PacketAnalyzer {
  analyzePacket(packet) {
    if (!packet.payload?.payload) return null;

    const ip = packet.payload.payload;
    const timestamp = new Date().toISOString();
    
    const packetInfo = {
      timestamp,
      sourceIp: ip.saddr.toString(),
      destIp: ip.daddr.toString(),
      protocol: ip.protocol,
      length: packet.pcap_header.len
    };

    if (ip.payload) {
      packetInfo.sourcePort = ip.payload.sport;
      packetInfo.destPort = ip.payload.dport;
      
      if (ip.payload.flags) {
        packetInfo.flags = ip.payload.flags;
      }
    }

    return packetInfo;
  }

  formatPacketInfo(packetInfo) {
    if (!packetInfo) return '';

    const protocolMap = {
      6: 'TCP',
      17: 'UDP'
    };

    const protocol = protocolMap[packetInfo.protocol] || packetInfo.protocol;
    
    let output = [
      chalk.yellow(packetInfo.timestamp),
      '|',
      chalk.green(`${packetInfo.sourceIp}:${packetInfo.sourcePort || '?'}`),
      '→',
      chalk.red(`${packetInfo.destIp}:${packetInfo.destPort || '?'}`),
      '|',
      chalk.blue(protocol),
      '|',
      `Length: ${packetInfo.length} bytes`
    ];

    if (packetInfo.flags) {
      output.push(`| Flags: ${packetInfo.flags}`);
    }

    return output.join(' ');
  }
}

module.exports = { PacketAnalyzer };